src/ServiceControl.Persistence.RavenDb5/RavenAttachmentsBodyStorage.cs

```diff
@@ -1,43 +1,61 @@
 ﻿namespace ServiceControl.Operations.BodyStorage.RavenAttachments
 {
+    using System;
     using System.IO;
+    using System.Linq;
     using System.Threading.Tasks;
     using Raven.Client.Documents;
-    using Raven.Client.Documents.Commands.Batches;
-    using Sparrow.Json.Parsing;
+    using Raven.Client.Documents.Session;
+    using ServiceControl.MessageFailures;
+    using ServiceControl.MessageFailures.Api;
 
     class RavenAttachmentsBodyStorage : IBodyStorage
     {
-        const string AttachmentName = "body";
+        public const string AttachmentName = "body";
         readonly IDocumentStore documentStore;
 
         public RavenAttachmentsBodyStorage(IDocumentStore documentStore)
         {
             this.documentStore = documentStore;
         }
 
-        // TODO: This method is only used in tests and not by ServiceControl itself! But in the Raven3.5 persister, it IS used!
-        // It should probably be removed and tests should use the RavenDbRecoverabilityIngestionUnitOfWork
-        public async Task Store(string messageId, string contentType, int bodySize, Stream bodyStream)
+        public Task Store(string uniqueId, string contentType, int bodySize, Stream bodyStream)
+            => throw new NotImplementedException("Only included for interface compatibility with Raven3.5 persister implementation. Raven5 tests should use IIngestionUnitOfWorkFactory to store failed messages/bodies.");
+
+        public async Task<MessageBodyStreamResult> TryFetch(string bodyId)
         {
-            var documentId = MessageBodyIdGenerator.MakeDocumentId(messageId);
+            using var session = documentStore.OpenAsyncSession();
+
+            // BodyId could be a MessageID or a UniqueID, but if a UniqueID then it will be a DeterministicGuid of MessageID and endpoint name and be Guid-parseable
```

Comment:
Why can it be a message ID or unique ID?

Reply:
Because in my redesign, we're always storing it using the unique ID, and the message metadata will say "look at `messages/{uniqueId}/body` for the body" so in theory, it should always be the unique ID. But previous versions stored the body only by message ID and while in theory, Pulse and Insight should pay attention to the message metadata that points to the uniqueID-based URL, it's impossible for me to say that assumption hasn't leaked into one of those apps. So just following what the API looks like, it's entirely reasonable some client somewhere could try to access by message ID.

Reply:
If we test SI + SP and they use the BodyUrl we could simplify this and only use Unique ID?